src/main/scala/resource/SharedResource.scala

```diff
@@ -0,0 +1,59 @@
+package resource
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.atomic.AtomicInteger
+
+final class SharedResource[A : Resource : Manifest](opener: => A) extends AbstractManagedResource[A] {
+	private val resource = implicitly[Resource[A]]
+	// A referencing counting atomic variable we use to "guard" this resource.
+	// Currently, we spin-lock accessing this ref because we don't expect opening to block a thread for very long.
+	private val ref = new AtomicReference[Option[A]](None)
+	private val count = new AtomicInteger(0)
+
+	protected def open: A = {
+		val myCount = count.addAndGet(1)
+		// If we win the race, we construct the resource
+		if (myCount == 1) {
+			// We must construct the reference
+			// We may want to CAS here and fail if we're inconsistent.
+			val r = opener
+			ref.lazySet(Some(r))
+			resource open r
+		}
+		// Lazy spin to get the current value
+		def get(): A = 
+		   ref.get match {
+		     case Some(r) => r
+		     case None => 
+		       // TODO - backoff strategy
+		       Thread.`yield`()
+		       get()
+		   }
+		get()
+	}
+	protected def unsafeClose(handle: A, errors: Option[Throwable]): Unit = {
+		val cnt = count.decrementAndGet()
+		// TODO - Throw error if cnt reaches negative?
+		if (cnt == 0) {
+			// Mark the resource as closed, while we clean up behind the scenes.
+			// If the CAS fails, it means someone else has already opened the resource again.
+			ref.compareAndSet(Some(handle), None)
```

Comment:
`compareAndSet` checks object reference equality, which is `eq` in Scala.
A newly created `Some` instance always does not `eq` to the one in `ref`.

We should store `null` or the `handle` itself, instead of an `Option`.